Mathlib/GroupTheory/Submonoid/Units.lean

```diff
@@ -0,0 +1,322 @@
+/-
+Copyright (c) 2023 Wrenna Robson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Wrenna Robson
+-/
+import Mathlib.GroupTheory.Submonoid.Operations
+import Mathlib.GroupTheory.Submonoid.Pointwise
+import Mathlib.GroupTheory.Subgroup.Basic
+
+/-!
+
+# Submonoid of units
+
+Given a submonoid `S` of a monoid `M`, we define the submonoid `S.units` as the units of `S` as a
+subgroup of `Mˣ`. That is to say, `S.units` contains all members of `S` which have a
+two-sided inverse within `S`, as terms of type `Mˣ`.
+This is multiplicatively equivalent to `Sˣ` and also to `IsUnit.submonoid S`,
+but it is as a `Subgroup Mˣ` rather than as a type in and of itself or a `Submonoid S`.
+
+-/
+
+variable {M : Type*}
+
+variable [Monoid M]
+
+open Pointwise in
+/-- The units of `S`, packaged as a subgroup of `Mˣ`.  -/
+@[to_additive " The additive units of `S`, packaged as an additive subgroup of `AddUnits M`. "]
+def Submonoid.units (S : Submonoid M) : Subgroup Mˣ where
+  toSubmonoid := S.comap (Units.coeHom M) ⊓ (S.comap (Units.coeHom M))⁻¹
+  inv_mem' ha := ⟨ha.2, ha.1⟩
+
+/-- A subgroup of units represented as a submonoid of `M`.  -/
+@[to_additive " A additive subgroup of additive units represented as a additive submonoid of `M`. "]
+def Subgroup.ofUnits (S : Subgroup Mˣ) : Submonoid M := S.toSubmonoid.map (Units.coeHom M)
+
+@[to_additive]
+lemma Submonoid.units_mono : Monotone (Submonoid.units (M := M)) :=
+fun _ _ hST _ ⟨h₁, h₂⟩ => ⟨hST h₁, hST h₂⟩
+
+@[to_additive (attr := simp)]
+lemma Submonoid.ofunits_units_le (S : Submonoid M) : S.units.ofUnits ≤ S :=
+fun  _ ⟨_, hm, he⟩ => he ▸ hm.1
+
+@[to_additive]
+lemma Subgroup.ofUnits_mono : Monotone (Subgroup.ofUnits (M := M)) :=
+fun _ _ hST _ ⟨x, hx, hy⟩ => ⟨x, hST hx, hy⟩
+
+@[to_additive (attr := simp)]
+lemma Subgroup.ofUnits_units_eq (S : Subgroup Mˣ) : S.ofUnits.units = S := Subgroup.ext (fun _ =>
+  ⟨fun ⟨⟨_, hm, he⟩, _⟩ => (Units.ext he) ▸ hm, fun hm => ⟨⟨_, hm, rfl⟩, ⟨_, S.inv_mem hm, rfl⟩⟩⟩)
+
+/-- A Galois coinsertion exists between the coercion from a subgroup of units to a submonoid and
+the reduction from a submonoid to its unit group. -/
+@[to_additive " A Galois coinsertion exists between the coercion from a additive subgroup of
+additive units to a additive submonoid and the reduction from a additive submonoid to its unit
+group. " ]
+def ofUnits_units_gci : GaloisCoinsertion (Subgroup.ofUnits) (Submonoid.units (M := M)) :=
+  GaloisCoinsertion.monotoneIntro Submonoid.units_mono Subgroup.ofUnits_mono
+  Submonoid.ofunits_units_le Subgroup.ofUnits_units_eq
+
+@[to_additive]
+lemma ofUnits_units_gc : GaloisConnection (Subgroup.ofUnits) (Submonoid.units (M := M)) :=
+ofUnits_units_gci.gc
+
+namespace Submonoid
+
+section Units
+
+@[to_additive (attr := simp)]
+lemma mem_units_iff (S : Submonoid M) (x : Mˣ) : x ∈ S.units ↔
```

Comment:
Maybe reverse the two sides? The RHS seems more complicated than the LHS (asking since it is `simp`).

Reply:
Perhaps it shouldn't be simp.

Reply:
Removed simp attribute.